Src/Newtonsoft.Json/Linq/JValue.Async.cs

```diff
@@ -0,0 +1,138 @@
+// Software is furnished to do so, subject to the following
+// conditions:
+//
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+// OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+// WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+// OTHER DEALINGS IN THE SOFTWARE.
+#endregion
+
+#if !(NET20 || NET35 || NET40 || PORTABLE40)
+
+using System;
+using System.Globalization;
+#if !PORTABLE || NETSTANDARD1_1
+using System.Numerics;
+#endif
+using System.Threading;
+using System.Threading.Tasks;
+using Newtonsoft.Json.Utilities;
+
+namespace Newtonsoft.Json.Linq
+{
+    public partial class JValue
+    {
+        /// <summary>
+        /// Writes this token to a <see cref="JsonWriter"/> asynchronously.
+        /// </summary>
+        /// <param name="writer">A <see cref="JsonWriter"/> into which this method will write.</param>
+        /// <param name="cancellationToken">The token to monitor for cancellation requests.</param>
+        /// <param name="converters">A collection of <see cref="JsonConverter"/> which will be used when writing the token.</param>
+        /// <returns>A <see cref="Task"/> that represents the asynchronous write operation.</returns>
+        public override Task WriteToAsync(JsonWriter writer, CancellationToken cancellationToken, params JsonConverter[] converters)
+        {
+            if (converters != null && converters.Length > 0 && _value != null)
+            {
+                JsonConverter matchingConverter = JsonSerializer.GetMatchingConverter(converters, _value.GetType());
+                if (matchingConverter != null && matchingConverter.CanWrite)
+                {
+                    // TODO: Call WriteJsonAsync when it exists.
+                    matchingConverter.WriteJson(writer, _value, JsonSerializer.CreateDefault());
+                    return AsyncUtils.CompletedTask;
+                }
+            }
+
+            switch (_valueType)
+            {
+                case JTokenType.Comment:
+                    return writer.WriteCommentAsync(_value != null ? _value.ToString() : null, cancellationToken);
+                case JTokenType.Raw:
+                    return writer.WriteRawValueAsync(_value != null ? _value.ToString() : null, cancellationToken);
+                case JTokenType.Null:
+                    return writer.WriteNullAsync(cancellationToken);
+                case JTokenType.Undefined:
+                    return writer.WriteUndefinedAsync(cancellationToken);
+                case JTokenType.Integer:
+                    if (_value is int)
```

Comment:
Do you mean doing as-and-null-check instead of is-and-cast? All the types involved here are value types, so that can't be done.